src/module/elements.rs

```diff
@@ -0,0 +1,185 @@
+//! Table elements within a wasm module.
+
+use crate::error::Result;
+use crate::module::emit::{Emit, IdsToIndices};
+use crate::module::functions::FunctionId;
+use crate::module::tables::TableId;
+use crate::module::Module;
+use crate::passes::Used;
+use id_arena::{Arena, Id};
+use parity_wasm::elements;
+use std::collections::HashMap;
+
+/// A passive element segment identifier
+pub type ElementId = Id<Element>;
+
+/// A passive segment which contains a list of functions
+#[derive(Debug)]
+pub struct Element {
+    members: Vec<FunctionId>,
+}
+
+/// All element segments of a wasm module, used to initialize `anyfunc` tables,
+/// used as function pointers.
+#[derive(Debug, Default)]
+pub struct ModuleElements {
+    active: HashMap<TableId, Vec<Option<FunctionId>>>,
+    passive: Arena<Element>,
+    index_to_element_id: HashMap<u32, ElementId>,
+}
+
+impl ModuleElements {
+    /// Parses a raw was section into a fully-formed `ModuleElements` instance.
+    pub fn parse(module: &Module, section: &elements::ElementSection) -> Result<ModuleElements> {
+        let mut active = HashMap::new();
+        let mut passive = Arena::new();
+        let mut index_to_element_id = HashMap::new();
+
+        for (i, segment) in section.entries().iter().enumerate() {
+            if segment.passive() {
+                let mut list = Vec::with_capacity(segment.members().len());
+                for &func in segment.members() {
+                    match module.funcs.function_for_index(func) {
+                        Some(id) => list.push(id),
+                        None => failure::bail!("invalid segment initialization"),
+                    }
+                }
+                let id = passive.next_id();
+                let ret = passive.alloc(Element { members: list });
+                debug_assert_eq!(id, ret);
+                index_to_element_id.insert(i as u32, id);
+                continue;
+            }
+
+            let table = module
+                .tables
+                .table_for_index(segment.index())
+                .ok_or_else(|| failure::format_err!("invalid table index"))?;
+            let list = active.entry(table).or_insert(Vec::new());
+            let offset = segment.offset().as_ref().unwrap();
+            if offset.code().len() != 2 {
+                failure::bail!("invalid initialization expression");
+            }
+            match offset.code()[1] {
+                elements::Instruction::End => {}
+                _ => failure::bail!("invalid initialization expression"),
+            }
+            let offset = match offset.code()[0] {
+                elements::Instruction::I32Const(n) => n as usize,
+                _ => failure::bail!("invalid initialization expression"),
```

Comment:
Maybe worth re-using our dummy local functions as init exprs stuff here (and cleaning it up some more). I expect that const expressions will be expanded and eventually include arbitrary arithmetic.

They can also use imported globals already, so this is technically incomplete (even though I haven't seen it in the wild yet).

Reply:
Yeah I figured that this'll need to be shored up one way or another, but I wasn't sure where it would best fit today. It's slightly different from the globals side as that was just validation as opposed to extracting the actual value. Do you think we should have like a `const_eval` function which evaluates to a `ConstValue` where one variant is a global perhaps?

Reply:
I'll open up a follow-up issue.